### ISSUE TITLE
Subscriptions with additional filters. Not only topic

### ISSUE THREAD

**Issue by AmbroiseCouissin** [ChilliCream/graphql-platform#5579]
### Is your feature request related to a problem?

Hello, 

I am currently using Subscriptions the following way:

```
    [Subscribe]
    [Topic("MESSAGE_SEND")]
    public Message OnSend(MessageWhere where, [EventMessage] Message publishedMessage) =>
        new List<Message> { publishedMessage }.AsQueryable().Filter(where).FirstOrDefault();
```

When the topic "MESSAGE_SEND" is sent from a `ITopicEventSender`, it reaches that function and the message gets filter with a very complicated `Filter(where). But if the message is filtered out, it will return null.

Is there any way for it to not return anything? Instead of null?

### The solution you'd like

Basically, when we reach a subscription, is there a way to be able to not send back anything to the client, instead of returning null?

### Product

Hot Chocolate

**Comment by aaulthudl** [comment]
Also interested in this. Apollo have functionality for filtering out subscriptions which I find lacking in Hot Chocolate 

https://www.apollographql.com/docs/apollo-server/data/subscriptions/#filtering-events

**Comment by jfloodnet** [comment]
Was just looking up how to implement this also.  Can someone provide some guidance on this? @AmbroiseCouissin is there anyway to achieve subscription filters as it is currently implemented?

**Comment by jfloodnet** [comment]
Ok I've been able to achieve this with ITopicEventReceiver

Here's an example of how to do it, essentially with read the events, filter them out and project to a new IAsyncEnumerable<EventUpdate> stream. 

This works, although I can't find any documentation that makes it obvious you can do this.

```
public class EventSubscription
{
    public async IAsyncEnumerable<EventUpdate> SubscribeToUpdates(
        [Service] ITopicEventReceiver receiver, int groupId)
    {
        var stream = await receiver.SubscribeAsync<EventUpdate>("eventUpdates");

        await foreach (var eventUpdate in stream .ReadEventsAsync())
        {
            if (eventUpdate.GroupId == groupId)
            {
                yield return eventUpdate;
            }
        }
    }

    [Subscribe(With = nameof(SubscribeToUpdates))]
    public EventUpdate OnEventUpdated([EventMessage] EventUpdate eventUpdate, int
```

**Comment by aaulthudl** [comment]
@jfloodnet 

The code you put in works great 🙌 

I'm unsure if your use-case was just for an example but you could more simply create a topic for each `groupId` and only send to that topic specifically.

Where your subscription is defined:
```csharp
  public static string OnUpdatesTopic(int groupId) => "eventUpdates" + groupId;

  public async IAsyncEnumerable<EventUpdate> SubscribeToUpdates([Service] ITopicEventReceiver receiver, int groupId) => receiver.SubscribeAsync< EventUpdate >(OnUpdatesTopic(groupId));
```

Where you trigger the subscription:
```csharp
await _sender.SendAsync(SubscriptionClassName.OnUpdatesTopic(groupId), updateContents);
```

**Comment by jfloodnet** [comment]
Thanks @aaulthudl, yes my example was for simplicity, I do intend to allow the user to provide more fine grain filtering params

**Comment by AmbroiseCouissin** [comment]
I think my problem is a bit more complex.

1. First, let's define a subscription operation:

```
    public async ValueTask<ISourceStream<NotificationMessage>> SubscribeToCreatedNotificationMessages(IResolverContext context, [Service] ITopicEventReceiver receiver)
    {
        await context.PopulateAuthParamsAsync();
        return await receiver.SubscribeAsync<NotificationMessage>(MessageEventType.MESSAGE_NOTIFICATION_CREATE.ToString());
    }
```

Then let's play with some saved auth stuff and get a tenantId, and also define a filter in the subscription to do fine tuned dynamic subscription:

```
    [Subscribe(With = nameof(SubscribeToCreatedNotificationMessages))]
    public NotificationMessage NotificationMessageCreated(IResolverContext context, NotificationMessageWhere where, [EventMessage] NotificationMessage publishedMessage)
    {
        AuthResponse authResponse = context.GetAuthParams();
        NotificationMessageWhere combinedWhere = new()
        {
            And = new List<NotificationMessageWhere>
            {
                new() { TenantId = new StringFilter { Eq = authResponse.TenantId } },
                where
            }
        };

        return new List<NotificationMessage> { publishedMessage }.AsQueryable().Filter(combinedWhere).FirstOrDefault();
    }
```

3. Then, let's publish an event that should trigger what's above:

```
...
ITopicEventSender eventSender = context.Service<ITopicEventSender>();
await eventSender.SendAsync(MessageEventType.MESSAGE_NOTIFICATION_CREATE.ToString(), result.Value /* the NotificationMessage published */, ct);
...
```



I want to be able to not send back any message to the subscribers if the filter should not take it into account.
Currently I send back `null`, but I want to not even send back anything.

Any idea on how to do that?

**Comment by jfloodnet** [comment]
Hi @AmbroiseCouissin 

At first glance it should be doable and the same principle should apply 

Something to note is that I perform all the filtering inside that first function `SubscribeToCreatedNotificationMessages`, my respective method to your NotificationMessageCreated is just a shell. So I'd try pushing your filtering inside SubscribeToCreatedNotificationMessages.

This is what my simplified version looked like 

```
    public async IAsyncEnumerable<EventUpdate> SubscribeToUpdates(
        [Service] ITopicEventReceiver receiver, int groupId)
    {
        var stream = await receiver.SubscribeAsync<EventUpdate>("eventUpdates");

        await foreach (var eventUpdate in stream.ReadEventsAsync())
        {
            if (eventUpdate.GroupId == groupId)
            {
                yield return eventUpdate;
            }
        }
    }
```
    
Applying the above pattern to your code:
    
 ```
    public async IAsyncEnumerable<NotificationMessage> SubscribeToCreatedNotificationMessages(IResolverContext context, [Service] ITopicEventReceiver receiver)
    {
        await context.PopulateAuthParamsAsync();
        var stream = await receiver.SubscribeAsync<NotificationMessage>(MessageEventType.MESSAGE_NOTIFICATION_CREATE.ToString());

        AuthResponse authResponse = context.GetAuthParams();
        NotificationMessageWhere combinedWhere = new()
        {
            And = new List<NotificationMessageWhere>
            {
                new() { TenantId = new StringFilter { Eq = authResponse.TenantId } },
                where
            }
        };

        var messages = await stream.ReadEventsAsync();

        var filteredMessages = messages.AsQueryable().Filter(combinedWhere);

        await foreach (var message in filteredMessages)
        {
            yield return message;
        }
    }
    
    [Subscribe(With = nameof(SubscribeToCreatedNotificationMessages))]
    public NotificationMessage NotificationMessageCreated(IResolverContext context, NotificationMessageWhere where, [EventMessage] NotificationMessage publishedMessage)
    {
        return publishedMessage;
    }
    
```

**Comment by AmbroiseCouissin** [comment]
Hello @jfloodnet 

Thanks for your help again! That's so helpful.

There are an issue with this approach:

The use of IAsyncEnumerable makes use unable to use IResolverContext correctly as ContextData is Disposed. That means I can't access the access token that is in the HttpContext that is in the ContextData, that is in the IResolverContext.

Any other idea?

**Comment by jfloodnet** [comment]
So I'm not sure where exactly I read it, but I had originally followed this article for setting up authorisation. https://chillicream.com/docs/hotchocolate/v12/security/authorization

I can't see it in there, but I found that you can simple accept the ClaimsPrincipal into your Subsciber function.

```
   public async IAsyncEnumerable<NotificationMessage> SubscribeToCreatedNotificationMessages([Service] ITopicEventReceiver receiver, ClaimsPrincipal claimsPrincipal)

```

Then it's just a matter for you to ensure that the users claim contains the tenantId and you should be good.

**Comment by AmbroiseCouissin** [comment]
I solved this by not using the IResolvedContext has this is disposed on IAsyncEnumerable. All the rest seens to work!

Thanks for your help!

**Comment by aaulthudl** [comment]
I feel this is still valid as a feature request. It would be good if we could have something like a `filter` attribute which allows you to skip subscription triggers.

**Comment by andreyleskov** [comment]
Thanks a lot guys! Stumled upon same problem today and found the solution here:

> Ok I've been able to achieve this with ITopicEventReceiver
> 
> Here's an example of how to do it, essentially using ReadEventsAsync, filter them out and project to a new IAsyncEnumerable stream.
> 
> This works, although I can't find any documentation that makes it obvious you can do this.
> 
> ```
> public class EventSubscription
> {
>     public async IAsyncEnumerable<EventUpdate> SubscribeToUpdates(
>         [Service] ITopicEventReceiver receiver, int groupId)
>     {
>         var stream = await receiver.SubscribeAsync<EventUpdate>("eventUpdates");
> 
>         await foreach (var eventUpdate in stream.ReadEventsAsync())
>         {
>             if (eventUpdate.GroupId == groupId)
>             {
>                 yield return eventUpdate;
>             }
>         }
>     }
> 
>     [Subscribe(With = nameof(SubscribeToUpdates))]
>     public EventUpdate OnEventUpdated([EventMessage] EventUpdate eventUpdate, int groupId) => eventUpdate;
>     
> }
>     
> ```

I also use JWT for auth, but permissions are a bit simpler then for @AmbroiseCouissin and I'm to check it only at the moment of the subscription.
